### PR TITLE
Improve readability of `impl/data.rb`

### PR DESCRIPTION
**Pull request by ashmaroli**
Improve readability of public methods in `impl/data.rb` by abstracting nested `if..else` to private methods

### CODE REVIEW

File: lib/wab/impl/data.rb [hunk 1]
```diff
@@ -37,27 +38,14 @@ def initialize(value, repair, check=true)
       # same as ['child', 'grandchild'].
       def get(path)
         if path.is_a?(Symbol)
-          node = @root[path] 
+          node = root[path]
         else
```
Comment (ohler55) [inline lib/wab/impl/data.rb]: Why the change? @root is faster access I believe.

File: lib/wab/impl/data.rb [hunk 2]
```diff
@@ -83,73 +71,51 @@ def set(path, value, repair=false)
...
-              node = nn
-            end
-          else
-            raise StandardError.new("Can not set a member of an #{node.class}.")
-          end
-        }
+        node = assign_node(path, root)
+
         key = path[-1]
+        ensure_hash_or_array(node)
```
Comment (ohler55) [inline lib/wab/impl/data.rb]: With this approach the check for Hash and Array is done twice. Better to check only once.

File: lib/wab/impl/data.rb [hunk 3]
```diff
@@ -83,73 +71,51 @@ def set(path, value, repair=false)
...
-        c.instance_variable_set(:@root, clone_value(@root))
+        c.instance_variable_set(:@root, clone_value(root))
         c
       end
 
       # Returns the instance converted to native Ruby values such as a Hash,
       # Array, etc.
       def native()
-        @root
+        root
```
Comment (ohler55) [inline lib/wab/impl/data.rb]: Same comment as earlier about efficiency of @root vs root method. If using the root method then native should be an alias instead. That's probably the correct approach, an alias.

File: lib/wab/impl/data.rb [hunk 4]
```diff
@@ -37,27 +38,14 @@ def initialize(value, repair, check=true)
...
-              if 0 == i && '0' != key && 0 != key
-                node = nil
-                break
-              end
-              node = node[i]
-            else
-              node = nil
-              break
-            end
-          }
```
Comment (ohler55) [inline lib/wab/impl/data.rb]: I'm not seeing the advantage here. The extract_node method is only called from this function which now drops to 7 lines with the build in extract_node. Not much of an improvement as now the reader needs to bounce down to the bottom of the file to see what is going on. I would see it as an improvement if any other method used extract_node but there is none.

File: lib/wab/impl/data.rb [hunk 5]
```diff
@@ -83,73 +71,51 @@ def set(path, value, repair=false)
...
-              else
-                node[key] = nn
-              end
-              node = nn
-            end
-          else
-            raise StandardError.new("Can not set a member of an #{node.class}.")
-          end
-        }
+        node = assign_node(path, root)
```
Comment (ohler55) [inline lib/wab/impl/data.rb]: Similar comment as extract_node. assign_node is not used anywhere else and only cause a discontinuity when reading the code.

File: lib/wab/impl/data.rb (hunk 1 above)
Comment (ashmaroli) [inline lib/wab/impl/data.rb]: You'll then able to `read` the value of `Data.root` outside `WAB::Impl::Data`. If there won't be a need for that, then, I'll revert this..

File: lib/wab/impl/data.rb (hunk 2 above)
Comment (ashmaroli) [inline lib/wab/impl/data.rb]: hmm.. I just moved it to a private helper method.
the `raise unless ..` was being called twice anyways..

File: lib/wab/impl/data.rb (hunk 3 above)
Comment (ashmaroli) [inline lib/wab/impl/data.rb]: Yes. I wanted to gauge the response for the `root` method first.
And then use `alias_method` instead of `alias`.. but if you feel `alias` is the better bet here, I'm fine with that.

File: lib/wab/impl/data.rb (hunk 2 above)
Comment (ohler55) [inline lib/wab/impl/data.rb]: Raising an error situation so need for performance there. Checking is_a? Array or Hash twice puts that check on every call.

File: lib/wab/impl/data.rb (hunk 3 above)
Comment (ohler55) [inline lib/wab/impl/data.rb]: Using an alias instead of calling the root method would have to be faster just since it reduces one call. @root vs root timing I'm not sure of. There might be some optimization somewhere for that but at least in older versions of Ruby @root would have been faster.

Another plus for using an alias is that it is just one line. Not that important though.

File: lib/wab/impl/data.rb (hunk 2 above)
Comment (ashmaroli) [inline lib/wab/impl/data.rb]: okay.. say, we move checking `is_a? Array or Hash` to an utility method..
would `raise "something" unless valid_datatype` be better?

**Comment by ashmaroli** [comment]
Okay. Let's defer changing `@root` to `root` for future.
Is there anything worth merging in this PR? I'll rebase with updates or close otherwise

**Comment by ohler55** [comment]
If you would like to make that alias you can. Maybe a separate PR would be easier than reverting though.
